llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp

```diff
@@ -434,18 +455,8 @@ static void createMemMoveLoopUnknownSize(Instruction *InsertBefore,
   // If the pointers are in different address spaces, they need to be converted
   // to a compatible one. Cases where memory ranges in the different address
   // spaces cannot overlap are lowered as memcpy and not handled here.
-  Value *CmpSrcAddr = SrcAddr;
-  Value *CmpDstAddr = DstAddr;
-  if ((SrcAS != DstAS)) {
-    if (TTI.isValidAddrSpaceCast(DstAS, SrcAS)) {
-      CmpDstAddr = PLBuilder.CreateAddrSpaceCast(DstAddr, SrcAddr->getType());
-    } else if (TTI.isValidAddrSpaceCast(SrcAS, DstAS)) {
-      CmpSrcAddr = PLBuilder.CreateAddrSpaceCast(SrcAddr, DstAddr->getType());
-    } else {
-      llvm_unreachable("Can only lower memmove between address spaces if they "
-                       "support addrspacecast");
-    }
-  }
+  auto [CmpSrcAddr, CmpDstAddr] =
+      tryInsertCastToCommonAddrSpace(PLBuilder, SrcAddr, DstAddr, TTI);
```

Comment:
I don't see how this bails out if the cast isn't possible. It will go ahead to create the invalid compare with different pointer types?

Reply:
I think I misunderstood what you meant with ignoring unhandled cases; did you mean that the `createMemMoveLoop*` functions should do nothing and communicate it to the caller if the cast is impossible?
(Note that this would just be future-proofing; in the current state, the `createMemMoveLoop*` functions are only called in `llvm::expandMemMoveAsLoop`, after a check that rules out impossible casts [here](https://github.com/ritter-x2a/llvm-project/blob/wide-memmove/llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp#L938))

Reply:
OK, if it's pre-checked the unreachable before was fine

Reply:
Restored in a082df5.